File: cpp/scripts/randomtest.py
import random
import time

# Running the test script to stress test the matching engine, this does not save failed tests:
#   python3 scripts/randomtest.py | ./grader engine
# To save tests that failed, refer to `randomtest.sh`

def main():
    ticker_symbols = ["META", "AAPL", "NFLX", "GOOGL", "AMZN", "MSFT", "TSLA", "NVDA", "INTC", "AMD"]
    num_of_threads=50

    # first line
    print(str(num_of_threads) + "")
    print("0-" + str(num_of_threads - 1) + " o")
    order_id = 1
    for i in range(num_of_threads):
        random.seed(time.time())
        print(f"{i} B {order_id} {random.choice(ticker_symbols)} 2705 30")
        order_id += 1
        print(f"{i} S {order_id} {random.choice(ticker_symbols)} 3260 1")
        order_id += 1
        print(f"{i} C {order_id - 2}")
        print(f"{i} S {order_id} {random.choice(ticker_symbols)} 2701 25")
        order_id += 1
        print(f"{i} B {order_id} {random.choice(ticker_symbols)} 3290 10")
        order_id += 1
        print(f"{i} C {order_id - 3}")

    # last 2 lines
    print(".")
    print("0-" + str(num_of_threads - 1) + " x")
    
if __name__ == "__main__":
    main()

File: go/scripts/randomtest.go
package main

import (
	"flag"
	"fmt"
	"math/rand"
)

// Usage: go run randomtest.go | ./grader engine

func main() {
	var clients int
	flag.IntVar(&clients, "c", 50, "number of concurrent clients")
	flag.Parse()

	tickerSymbols := []string{
		"META", "AAPL", "NFLX", "GOOGL", "AMZN", "MSFT", "TSLA", "NVDA", "INTC", "AMD",
	}
	randomSymbol := func() string { return tickerSymbols[rand.Intn(len(tickerSymbols))] }

	prices := []int{2705, 3260, 2701, 3290}
	randomPrice := func() int { return prices[rand.Intn(len(prices))] }

	count := []int{10, 25, 30, 50}
	randomCount := func() int { return count[rand.Intn(len(count))] }

	// first line
	fmt.Println(clients)
	fmt.Printf("0-%d o\n", clients-1)
	orderID := 1
	for i := 0; i < clients; i++ {
		fmt.Printf("%v B %v %v %v %v\n", i, orderID, randomSymbol(), randomPrice(), randomCount())
		orderID++
		fmt.Printf("%v S %v %v %v %v\n", i, orderID, randomSymbol(), randomPrice(), randomCount())
		orderID++
		fmt.Printf("%v C %v\n", i, orderID-2)
		fmt.Printf("%v S %v %v %v %v\n", i, orderID, randomSymbol(), randomPrice(), randomCount())
		orderID++
		fmt.Printf("%v B %v %v %v %v\n", i, orderID, randomSymbol(), randomPrice(), randomCount())
		orderID++
		fmt.Printf("%v C %v\n", i, orderID-3)
	}

	// last 2 lines
	fmt.Println(".")
	fmt.Printf("0-%d x\n", clients-1)
}
